wpilibj/src/athena/java/edu/wpi/first/wpilibj/QuantityEncoder.java

```diff
@@ -0,0 +1,80 @@
+package edu.wpi.first.wpilibj;
+
+import tec.uom.se.quantity.Quantities;
+
+import javax.measure.Quantity;
+import javax.measure.Unit;
+
+/**
+ *
+ * @param <Q> The Quantity being measured. (ex: Length, Angle)
+ * @param <S> The Speed of the Quantity being measured. (ex: Speed, AngularSpeed)
+ */
+// Package private so it is not part of the public API
+abstract class QuantityEncoder<Q extends Quantity<Q>, S extends Quantity<S>> {
+
+  private final Encoder m_encoder;
+
+  public QuantityEncoder(int channelA, int channelB, Quantity<Q> quantityPerPulse) {
+    m_encoder = new Encoder(channelA, channelB);
+
+    m_encoder.setDistancePerPulse(quantityPerPulse.to(quantityPerPulse.getUnit().getSystemUnit())
+        .getValue().doubleValue());
+  }
+
+  public QuantityEncoder(int channelA, int channelB, int indexChannel,
+                      Quantity<Q> quantityPerPulse) {
+    m_encoder = new Encoder(channelA, channelB, indexChannel);
+
+    m_encoder.setDistancePerPulse(quantityPerPulse.to(quantityPerPulse.getUnit().getSystemUnit())
+        .getValue().doubleValue());
+  }
+
+  public QuantityEncoder(DigitalSource sourceA, DigitalSource sourceB,
+                      Quantity<Q> quantityPerPulse) {
+    m_encoder = new Encoder(sourceA, sourceB);
+
+    m_encoder.setDistancePerPulse(quantityPerPulse.to(quantityPerPulse.getUnit().getSystemUnit())
+        .getValue().doubleValue());
+  }
+
+  public QuantityEncoder(DigitalSource sourceA, DigitalSource sourceB, DigitalSource indexSource,
+                      Quantity<Q> quantityPerPulse) {
+    m_encoder = new Encoder(sourceA, sourceB, indexSource);
+
+    m_encoder.setDistancePerPulse(quantityPerPulse.to(quantityPerPulse.getUnit().getSystemUnit())
+        .getValue().doubleValue());
+  }
+
+  /**
+   * Set the minimum rate of the device before the hardware reports it stopped.
+   *
+   * @param minRate The minimum speed.
+   */
+  public void setMinimumSpeed(Quantity<S> minRate) {
+    m_encoder.setMinRate(minRate.to(minRate.getUnit().getSystemUnit()).getValue().doubleValue());
+  }
+
+  /**
+   * Get the speed of the encoder.
+   *
+   * @return The speed of the encoder.
+   */
+  public Quantity<S> getSpeed() {
+    return Quantities.getQuantity(m_encoder.getRate(), getSpeedUnit().getSystemUnit());
+  }
+
+  /**
+   * Get the distance the encoder has driven.
+   *
+   * @return The distance driven since the last reset
+   */
+  public Quantity<Q> get() {
+    return Quantities.getQuantity(m_encoder.getDistance(), getUnit().getSystemUnit());
+  }
+
+  protected abstract Unit<Q> getUnit();
+
+  protected abstract Unit<S> getSpeedUnit();
```

Comment:
If I change these to field, then I cannot override the value in Distance and Angle encoder.

I think it is a good idea to change these to be public methods so users can access them.  I probably need to come up with a better method name.  `getNativeUnit()`?